Move Package.Dependency.Constraint to Dune_lang

gridbugs opened the pull request:
This will let us use the constraint language within lockfiles which we need in order for build and install commands be conditional on opam filters.

gridbugs committed:
Move Package.Dependency.Constraint to Dune_lang

This will let us use the constraint language within lockfiles which we
need in order for build and install commands be conditional on opam
filters.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

Alizter commented on src/dune_lang/package_constraint.mli:
```diff
@@ -0,0 +1,33 @@
+            excluded from the string) *)
+end
+
+type t =
+  | Bvar of Var.t
+  | Uop of Op.t * Var.t
+  | Bop of Op.t * Var.t * Var.t
+  | And of t list
+  | Or of t list
```
Can you add comments for these?

gridbugs replied:
Done

Leonidas-from-XIV reviewed (commented):
I think this PR makes a good deal of sense, some comments that popped into my head while I was reviewing below.

gridbugs commented on src/dune_lang/package_constraint.ml:
```diff
@@ -0,0 +1,121 @@
+  let decode =
+    let open Dune_sexp.Decoder in
+    let+ s = string in
+    if String.is_prefix s ~prefix:":" then Var (String.drop s 1) else Literal s
```
I'll do that in a separate PR

gridbugs commented on src/dune_rules/package.mli:
```diff
@@ -33,27 +33,27 @@ module Id : sig
-
   module Constraint : sig
+    module Op : sig
+      type t = Dune_lang.Package_constraint.Op.t =
```
also `Var.t`

gridbugs committed:
Clarify handling of values in package constraints

This restricts the argument to the `Bvar` constructor to variables.
Previously string literals were also permitted in this position though
the parser never produced such values. Also adds some documentation
about the constraint language.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs commented:
In the process of documenting the different `Constraint.t` constructors I noticed it was possible to have a `Bvar` which contained a string literal. Even though the parser was incapable of producing such a value, the fact that it could be represented made the API harder for me to understand. I've changed it so that only variables can be passed to the `Bvar` constructor and made a clear distinction between the `Variable.t` and `Value.t` types (previously string literals were in a type called `Var.t` which made them seem like variables and I've always found that confusing).

rgrinberg commented:
The change in the description looks fine, but I see you've started pushing some additional work on top. Was that intentional?

Alizter commented on src/dune_lang/package_constraint.mli:
```diff
@@ -0,0 +1,47 @@
+    operator which returns a boolean. Variables are dynamically typed and can
+    represent strings and booleans. If a variable appears in a position where a
+    boolean is expected it will be assumed to represent a boolean. *)
+type t =
```
Dumb question: Is our blang really that different from this?

gridbugs replied:
Good question. I don't know off the top of my head. Of course they both have similar functions though the package constraint language has the added requirement of being translatable to the opam filter language. At some point soon I'll be looking at actually evaluating conditions (currently we only use these to generate opam files) and at that point I'll see whether it makes sense to just replace this with blang.

gridbugs commented:
> The change in the description looks fine, but I see you've started pushing some additional work on top. Was that intentional?

Oops I accidentally pushed some additional work here. I just removed it.
